File: packages/ios/Sources/CivUI/CivButton.swift
// CivUI — CivButton for SwiftUI
// Accessible button component following government design system patterns.
// Supports primary, secondary, tertiary, and danger variants.

import SwiftUI

/// Button variant determines the visual treatment.
public enum ButtonVariant: String, CaseIterable {
    case primary, secondary, tertiary, danger
}

/// Accessible button for government applications.
///
/// Renders a styled button with variant-based colors. When `href` is set,
/// the button opens the URL as a link action. Mirrors the web `civ-button`
/// component.
///
/// VoiceOver announces the button label and disabled state.
///
/// Usage:
/// ```swift
/// CivButton(label: "Submit application", variant: .primary) {
///     submitForm()
/// }
/// ```
public struct CivButton: View {
    // MARK: - Properties

    /// Button text.
    public let label: String

    /// Visual variant (primary, secondary, tertiary, danger).
    public var emphasis: ButtonVariant

    /// Whether the button is disabled.
    public var isDisabled: Bool

    /// Icon name rendered before the label text.
    public var iconStart: String

    /// Icon name rendered after the label text.
    public var iconEnd: String

    /// When true, the label is treated as accessible name only; the button shows just the icon.
    public var isIconOnly: Bool

    /// When set, the button functions as a link action opening this URL.
    public var href: String?

    /// Called on button tap (parallels `click` event).
    public var onClick: (() -> Void)?

    /// Called for analytics tracking (parallels `civ-analytics` event).
    public var onAnalytics: ((String, [String: Any]?) -> Void)?

    /// Whether the button triggers a destructive action.
    public var danger: Bool

    /// Button type (e.g., "button", "submit", "reset").
    public var type: String

    /// Async-in-flight state. Swaps the leading icon for a spinner and
    /// disables the button. Should announce `loadingLabel` via VoiceOver.
    public var isLoading: Bool

    /// Accessible label announced via VoiceOver while `isLoading` is true.
    public var loadingLabel: String

    // MARK: - Internal State

    @Environment(\.colorScheme) private var colorScheme

    // MARK: - Initializer

    public init(
        label: String,
        emphasis: ButtonVariant = .primary,
        isDisabled: Bool = false,
        iconStart: String = "",
        iconEnd: String = "",
        isIconOnly: Bool = false,
        href: String? = nil,
        onClick: (() -> Void)? = nil,
        onAnalytics: ((String, [String: Any]?) -> Void)? = nil,
        danger: Bool = false,
        type: String = "button",
        isLoading: Bool = false,
        loadingLabel: String = "Loading…"
    ) {
        self.label = label
        self.emphasis = emphasis
        self.isDisabled = isDisabled
        self.iconStart = iconStart
        self.iconEnd = iconEnd
        self.isIconOnly = isIconOnly
        self.href = href
        self.onClick = onClick
        self.onAnalytics = onAnalytics
        self.danger = danger
        self.type = type
        self.isLoading = isLoading
        self.loadingLabel = loadingLabel
    }

    // MARK: - Body

    public var body: some View {
        Button(action: handleTap) {
            HStack(spacing: CivTokens.Spacing._1) {
                // When loading, replace the leading icon with a
                // ProgressView spinner. The label stays visible; the
                // accessibility name (below) swaps to `loadingLabel`.
                if isLoading {
                    ProgressView()
                        .progressViewStyle(.circular)
                        .scaleEffect(0.7)
                } else if !iconStart.isEmpty {
                    CivIcon(name: iconStart, size: "sm")
                }
                Text(label)
                if !iconEnd.isEmpty {
                    CivIcon(name: iconEnd, size: "sm")
                }
            }
                .font(.system(size: CivTokens.Typography.FontSize.base,
                              weight: CivTokens.Typography.FontWeight.semibold))
                .padding(.horizontal, CivTokens.Spacing._6)
                .padding(.vertical, CivTokens.Spacing._2_5)
                .foregroundColor(foregroundColor)
                .background(backgroundColor)
                .cornerRadius(CivTokens.Border.Radius.default_)
                .overlay(borderOverlay)
        }
        // Loading implies disabled so taps don't fire concurrent work.
        .disabled(isDisabled || isLoading)
        .opacity(isDisabled ? 0.5 : 1.0)
        // VoiceOver hears the loading label instead of the static
        // button label while loading is in flight.
        .accessibilityLabel(isLoading ? Text(loadingLabel) : Text(label))
        .accessibilityAddTraits(href != nil ? .isLink : .isButton)
        .accessibilityRemoveTraits(href != nil ? .isButton : .isLink)
    }

    // MARK: - Actions

    private func handleTap() {
        guard !isDisabled else { return }
        onAnalytics?("click", ["variant": variant.rawValue])

        if let href, let url = URL(string: href) {
            UIApplication.shared.open(url)
        }

        onClick?()
    }

    // MARK: - Variant Colors

    private var foregroundColor: Color {
        switch variant {
        case .primary:
            return adaptiveColor(light: CivTokens.Colors.White.default_,
                                 dark: CivTokens.DarkColors.White.default_)
        case .secondary:
            return adaptiveColor(light: CivTokens.Colors.Base.darkest,
                                 dark: CivTokens.DarkColors.Base.darkest)
        case .tertiary:
            return adaptiveColor(light: CivTokens.Colors.Primary.default_,
                                 dark: CivTokens.DarkColors.Primary.default_)
        case .danger:
            return adaptiveColor(light: CivTokens.Colors.White.default_,
                                 dark: CivTokens.DarkColors.White.default_)
        }
    }

    private var backgroundColor: Color {
        switch variant {
        case .primary:
            return adaptiveColor(light: CivTokens.Colors.Primary.default_,
                                 dark: CivTokens.DarkColors.Primary.default_)
        case .secondary:
            return adaptiveColor(light: CivTokens.Colors.Base.lightest,
                                 dark: CivTokens.DarkColors.Base.lightest)
        case .tertiary:
            return Color.clear
        case .danger:
            return adaptiveColor(light: CivTokens.Colors.Error.default_,
                                 dark: CivTokens.DarkColors.Error.default_)
        }
    }

    private var borderOverlay: some View {
        Group {
            if variant == .tertiary {
                RoundedRectangle(cornerRadius: CivTokens.Border.Radius.default_)
                    .stroke(
                        adaptiveColor(light: CivTokens.Colors.Primary.default_,
                                      dark: CivTokens.DarkColors.Primary.default_),
                        lineWidth: CivTokens.Border.Width._2
                    )
            } else {
                EmptyView()
            }
        }
    }

    // MARK: - Color Helper

    private func adaptiveColor(light: Color, dark: Color) -> Color {
        colorScheme == .dark ? dark : light
    }
}

// MARK: - Preview

#if DEBUG
struct CivButton_Previews: PreviewProvider {
    struct PreviewWrapper: View {
        var body: some View {
            ScrollView {
                VStack(alignment: .leading, spacing: CivTokens.Spacing._4) {
                    CivButton(label: "Submit application", variant: .primary) {
                        // Action
                    }

                    CivButton(label: "Save draft", variant: .secondary) {
                        // Action
                    }

                    CivButton(label: "Cancel", variant: .tertiary) {
                        // Action
                    }

                    CivButton(label: "Delete record", variant: .danger) {
                        // Action
                    }

                    CivButton(
                        label: "Visit agency website",
                        variant: .primary,
                        href: "https://www.usa.gov"
                    )

                    CivButton(
                        label: "Disabled button",
                        variant: .primary,
                        isDisabled: true
                    )
                }
                .padding()
            }
        }
    }

    static var previews: some View {
        PreviewWrapper()
            .previewDisplayName("Light")
        PreviewWrapper()
            .preferredColorScheme(.dark)
            .previewDisplayName("Dark")
    }
}
#endif

File: packages/android/src/main/kotlin/gov/civui/components/CivButton.kt
// CivUI — CivButton for Jetpack Compose
// Accessible button with emphasis styles. Renders as link when href is set.
// Mirrors the web `civ-button` component (Section 508 compliant)

package gov.civui.components

import androidx.compose.foundation.BorderStroke
import androidx.compose.foundation.isSystemInDarkTheme
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.text.ClickableText
import androidx.compose.material3.Button
import androidx.compose.material3.ButtonDefaults
import androidx.compose.material3.OutlinedButton
import androidx.compose.material3.Text
import androidx.compose.material3.TextButton
import androidx.compose.runtime.Composable
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.alpha
import androidx.compose.ui.platform.LocalUriHandler
import androidx.compose.ui.semantics.Role
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.semantics.role
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.text.AnnotatedString
import androidx.compose.ui.text.TextStyle
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextDecoration
import androidx.compose.ui.tooling.preview.Preview
import androidx.compose.ui.unit.dp
import gov.civui.tokens.CivTokens

/**
 * Accessible button for government applications.
 *
 * Supports four variants: primary (filled), secondary (outlined), tertiary
 * (text-only), and danger (filled red). When `href` is set, renders as a
 * tappable link instead of a button.
 *
 * TalkBack announces the label and role. Disabled state reduces opacity
 * to 0.5 and prevents interaction.
 *
 * Usage:
 * ```kotlin
 * CivButton(
 *     label = "Submit application",
 *     emphasis = "primary",
 *     onClick = { /* handle click */ },
 * )
 *
 * CivButton(
 *     label = "Learn more",
 *     emphasis = "tertiary",
 *     href = "https://www.usa.gov",
 * )
 * ```
 */
@Composable
fun CivButton(
    label: String,
    modifier: Modifier = Modifier,
    emphasis: String = "primary",
    danger: Boolean = false,
    type: String = "button",
    disabled: Boolean = false,
    iconStart: String = "",
    iconEnd: String = "",
    iconOnly: Boolean = false,
    href: String = "",
    loading: Boolean = false,
    loadingLabel: String = "Loading…",
    onClick: (() -> Unit)? = null,
    onAnalytics: ((event: String, data: Map<String, Any>?) -> Unit)? = null,
) {
    val isDark = isSystemInDarkTheme()

    val primaryColor = if (isDark) CivTokens.DarkColors.Primary.default_ else CivTokens.Colors.Primary.default_
    val whiteColor = if (isDark) CivTokens.DarkColors.White.default_ else CivTokens.Colors.White.default_
    val errorColor = if (isDark) CivTokens.DarkColors.Error.default_ else CivTokens.Colors.Error.default_

    // `loading` implies disabled so repeated taps don't fire concurrent
    // work. The effectiveDisabled cascade applies across all variants
    // below so each Material button respects both flags.
    val effectiveDisabled = disabled || loading

    val buttonModifier = modifier
        .alpha(if (effectiveDisabled) 0.5f else 1f)
        .semantics {
            // TalkBack hears the loading label instead of the static
            // button label while loading is in flight, matching the
            // web's aria-label swap.
            contentDescription = if (loading) loadingLabel else label
            role = Role.Button // Compose has no Role.Link; the href conditional was a no-op
        }

    val handleClick: () -> Unit = {
        // Guard against tap-while-loading at the handler level — the
        // enabled=!effectiveDisabled cascade below already blocks the
        // tap at the UI layer; this is belt-and-suspenders for any
        // path that calls handleClick programmatically.
        if (!loading) {
            onAnalytics?.invoke("click", mapOf("label" to label, "emphasis" to emphasis))
            onClick?.invoke()
        }
    }

    // Link mode — `loading` is ignored on links (matches the web's
    // `_isLoading = loading && !_isLink` rule). Use only `disabled`.
    if (href.isNotEmpty()) {
        val uriHandler = LocalUriHandler.current
        TextButton(
            onClick = {
                if (!disabled) {
                    handleClick()
                    uriHandler.openUri(href)
                }
            },
            enabled = !disabled,
            modifier = buttonModifier,
        ) {
            Text(
                text = label,
                style = TextStyle(
                    fontSize = CivTokens.Typography.FontSize.base,
                    fontWeight = FontWeight.SemiBold,
                    textDecoration = TextDecoration.Underline,
                ),
                color = primaryColor,
            )
        }
        return
    }

    // Button mode
    when (emphasis) {
        "secondary" -> {
            OutlinedButton(
                onClick = handleClick,
                enabled = !effectiveDisabled,
                border = BorderStroke(CivTokens.Border.Width._2, primaryColor),
                colors = ButtonDefaults.outlinedButtonColors(
                    contentColor = primaryColor,
                ),
                modifier = buttonModifier,
            ) {
                Text(
                    text = label,
                    style = TextStyle(
                        fontSize = CivTokens.Typography.FontSize.base,
                        fontWeight = FontWeight.SemiBold,
                    ),
                )
            }
        }
        "tertiary" -> {
            TextButton(
                onClick = handleClick,
                enabled = !effectiveDisabled,
                colors = ButtonDefaults.textButtonColors(
                    contentColor = primaryColor,
                ),
                modifier = buttonModifier,
            ) {
                Text(
                    text = label,
                    style = TextStyle(
                        fontSize = CivTokens.Typography.FontSize.base,
                        fontWeight = FontWeight.SemiBold,
                    ),
                )
            }
        }
        "danger" -> {
            Button(
                onClick = handleClick,
                enabled = !effectiveDisabled,
                colors = ButtonDefaults.buttonColors(
                    containerColor = errorColor,
                    contentColor = whiteColor,
                ),
                modifier = buttonModifier,
            ) {
                Text(
                    text = label,
                    style = TextStyle(
                        fontSize = CivTokens.Typography.FontSize.base,
                        fontWeight = FontWeight.SemiBold,
                    ),
                )
            }
        }
        else -> {
            // "primary" — default filled button
            Button(
                onClick = handleClick,
                enabled = !effectiveDisabled,
                colors = ButtonDefaults.buttonColors(
                    containerColor = primaryColor,
                    contentColor = whiteColor,
                ),
                modifier = buttonModifier,
            ) {
                Text(
                    text = label,
                    style = TextStyle(
                        fontSize = CivTokens.Typography.FontSize.base,
                        fontWeight = FontWeight.SemiBold,
                    ),
                )
            }
        }
    }
}

// MARK: - Preview

@Preview(showBackground = true, name = "CivButton")
@Composable
private fun CivButtonPreview() {
    Column(modifier = Modifier.padding(16.dp)) {
        CivButton(
            label = "Submit application",
            emphasis = "primary",
        )

        CivButton(
            label = "Save draft",
            emphasis = "secondary",
        )

        CivButton(
            label = "Cancel",
            emphasis = "tertiary",
        )

        CivButton(
            label = "Delete record",
            emphasis = "danger",
        )

        CivButton(
            label = "Disabled button",
            emphasis = "primary",
            disabled = true,
        )

        CivButton(
            label = "Visit USA.gov",
            emphasis = "tertiary",
            href = "https://www.usa.gov",
        )
    }
}
